OpenRA.Mods.Common/Traits/AffectsShroud.cs

```diff
@@ -153,23 +174,33 @@ void INotifyRemovedFromWorld.RemovedFromWorld(Actor self)
 				RemoveCellsFromPlayerShroud(self, p);
 		}
 
-		public virtual WDist Range => CachedTraitDisabled ? WDist.Zero : Info.Range;
+		public virtual WDist Range => IsTraitDisabled ? WDist.Zero : Info.Range;
 
-		void INotifyMoving.MovementTypeChanged(Actor self, MovementType type)
+		readonly struct Input : IEquatable<Input>
 		{
-			// Recalculate the visibility at our final stop position
-			if (type == MovementType.None && self.IsInWorld)
-			{
-				var centerPosition = self.CenterPosition;
-				var projectedPos = centerPosition - new WVec(0, centerPosition.Z, centerPosition.Z);
-				var projectedLocation = self.World.Map.CellContaining(projectedPos);
-				var pos = self.CenterPosition;
+			public readonly (CPos Cell, SubCell SubCell)[] FootprintOccupiedCells { get; }
+			public readonly WPos CenterPosition { get; }
 
-				cachedLocation = projectedLocation;
-				cachedPos = pos;
+			public Input((CPos Cell, SubCell SubCell)[] footprintOccupiedCells, WPos centerPosition)
+			{
+				FootprintOccupiedCells = footprintOccupiedCells;
+				CenterPosition = centerPosition;
+			}
 
-				UpdateShroudCells(self);
+			public bool Equals(Input other)
+			{
+				if (CenterPosition != other.CenterPosition) return false;
+				if (FootprintOccupiedCells == null ^ other.FootprintOccupiedCells == null) return false;
+				if (FootprintOccupiedCells == null && other.FootprintOccupiedCells == null) return true;
+				return FootprintOccupiedCells.SequenceEqual(other.FootprintOccupiedCells);
 			}
+
+			public override int GetHashCode() =>
+				HashCode.Combine(CenterPosition, FootprintOccupiedCells?[0]);
```

Comment:
this doesn't look right

Reply:
Really we only care about equality, I'm just adding a hash code for completeness. I'm aiming for a "good enough" hash code without worrying about it too much. Tweaked to `?.FirstOrDefault()` just in case of empty arrays though.